File: low-level-design/lld-coding/multi-language/01-basic-classes/bank_account.py
"""
Python Implementation: BankAccount Class
========================================
Features:
- Class-based OOP
- __init__ constructor
- Instance variables with self
- Private variables by convention (_balance)
- String formatting with f-strings
"""


class BankAccount:
    """Represents a bank account with basic operations"""

    def __init__(self, account_number, initial_balance=0):
        """
        Initialize a new bank account

        Args:
            account_number: Unique account identifier
            initial_balance: Starting balance (default: 0)
        """
        self.account_number = account_number
        self._balance = initial_balance  # _ prefix = private by convention
        print(f"Created account: {account_number}")

    def deposit(self, amount):
        """
        Deposit money into account

        Args:
            amount: Amount to deposit

        Returns:
            bool: True if successful, False otherwise
        """
        if amount <= 0:
            print("Error: Deposit amount must be positive")
            return False

        self._balance += amount
        print(f"Deposited ${amount:.2f}")
        return True

    def withdraw(self, amount):
        """
        Withdraw money from account

        Args:
            amount: Amount to withdraw

        Returns:
            bool: True if successful, False if insufficient funds
        """
        if amount <= 0:
            print("Error: Withdrawal amount must be positive")
            return False

        if amount > self._balance:
            print(f"Error: Insufficient funds. Balance: ${self._balance:.2f}")
            return False

        self._balance -= amount
        print(f"Withdrew ${amount:.2f}")
        return True

    def get_balance(self):
        """
        Get current account balance

        Returns:
            float: Current balance
        """
        return self._balance

    def __str__(self):
        """String representation for printing"""
        return f"Account {self.account_number}: ${self._balance:.2f}"

    def __repr__(self):
        """Developer-friendly representation"""
        return f"BankAccount('{self.account_number}', {self._balance})"


def main():
    """Demo the BankAccount class"""
    print("=" * 60)
    print("Python: Bank Account Example")
    print("=" * 60)

    # Create account
    account = BankAccount("ACC001", 1000)
    print(f"Initial balance: ${account.get_balance():.2f}\n")

    # Deposit
    account.deposit(500)
    print(f"After deposit of $500: ${account.get_balance():.2f}\n")

    # Withdraw
    account.withdraw(200)
    print(f"After withdrawal of $200: ${account.get_balance():.2f}\n")

    # Failed withdrawal
    account.withdraw(2000)
    print(f"Final balance: ${account.get_balance():.2f}\n")

    # String representation
    print("Using __str__:", account)
    print("Using __repr__:", repr(account))


if __name__ == "__main__":
    main()

File: low-level-design/lld-coding/multi-language/01-basic-classes/BankAccount.java
/*
 * Java Implementation: BankAccount Class
 * =======================================
 * Features:
 * - Classic class-based OOP
 * - Constructor with same name as class
 * - Private fields with getters
 * - Explicit types everywhere
 * - Strong encapsulation with private/public
 */

public class BankAccount {
    // Private instance variables
    private String accountNumber;
    private double balance;

    /**
     * Constructor - Creates a new bank account
     *
     * @param accountNumber Unique account identifier
     * @param initialBalance Starting balance
     */
    public BankAccount(String accountNumber, double initialBalance) {
        this.accountNumber = accountNumber;
        this.balance = initialBalance;
        System.out.println("Created account: " + accountNumber);
    }

    /**
     * Overloaded constructor with default balance of 0
     */
    public BankAccount(String accountNumber) {
        this(accountNumber, 0.0);
    }

    /**
     * Deposit money into account
     *
     * @param amount Amount to deposit
     * @return true if successful, false otherwise
     */
    public boolean deposit(double amount) {
        if (amount <= 0) {
            System.out.println("Error: Deposit amount must be positive");
            return false;
        }

        this.balance += amount;
        System.out.printf("Deposited $%.2f%n", amount);
        return true;
    }

    /**
     * Withdraw money from account
     *
     * @param amount Amount to withdraw
     * @return true if successful, false if insufficient funds
     */
    public boolean withdraw(double amount) {
        if (amount <= 0) {
            System.out.println("Error: Withdrawal amount must be positive");
            return false;
        }

        if (amount > this.balance) {
            System.out.printf("Error: Insufficient funds. Balance: $%.2f%n", this.balance);
            return false;
        }

        this.balance -= amount;
        System.out.printf("Withdrew $%.2f%n", amount);
        return true;
    }

    /**
     * Get current account balance
     *
     * @return Current balance
     */
    public double getBalance() {
        return this.balance;
    }

    /**
     * Get account number
     *
     * @return Account number
     */
    public String getAccountNumber() {
        return this.accountNumber;
    }

    /**
     * String representation (like __str__ in Python)
     */
    @Override
    public String toString() {
        return String.format("Account %s: $%.2f", this.accountNumber, this.balance);
    }

    /**
     * Main method - Demo the BankAccount class
     */
    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("Java: Bank Account Example");
        System.out.println("============================================================");

        // Create account
        BankAccount account = new BankAccount("ACC001", 1000);
        System.out.printf("Initial balance: $%.2f%n%n", account.getBalance());

        // Deposit
        account.deposit(500);
        System.out.printf("After deposit of $500: $%.2f%n%n", account.getBalance());

        // Withdraw
        account.withdraw(200);
        System.out.printf("After withdrawal of $200: $%.2f%n%n", account.getBalance());

        // Failed withdrawal
        account.withdraw(2000);
        System.out.printf("Final balance: $%.2f%n%n", account.getBalance());

        // String representation
        System.out.println("Using toString(): " + account);
    }
}
